src/query/processor/cypher/semanticanalyzer/SemanticAnalyzer.cpp

```diff
@@ -0,0 +1,261 @@
+            {
+                string tempType;
+                if(child->elements[0]->nodeType == "VARIABLE")
+                {
+                    tempType = scopeManager->getType(child->elements[0]->value);
+                }else
+                {
+                    tempType = "ANY";
+                }
+                pair<string,string> x = pair<string,string>(child->elements[0]->value,tempType);
+                temp->insert(x);
+            } else if(child->nodeType == "VARIABLE")
+            {
+                string tempType = scopeManager->getType(child->value);
+                pair<string,string> x = pair<string,string>(child->value,tempType);
+                temp->insert(x);
+            }
+            else
+            {
+                reportError("use 'as' keyword to assign it to new variable" + node->value, node);
+                return false;
+            }
+        }
+        scopeManager->clearTable();
+        for (auto tempNode = temp->begin(); tempNode != temp->end(); ++tempNode) {
+            scopeManager->addSymbol(tempNode->first,tempNode->second);
+        }
+    }
+    else
+    {
+        for(int i=0; i<root->elements.size(); i++)
+        {
+            if(!analyze(root->elements[i]))
+            {
+                return false;
+            }
+        }
+    }
+    return true;
+}
+
+// Check for variable declarations
+bool SemanticAnalyzer::checkVariableDeclarations(ASTNode* node, string type) {
+     if (!scopeManager->lookup(node->value)){
+        // If variable is not in current scope, add it
+        scopeManager->addSymbol(node->value, type);
+    }else if(scopeManager->lookup(node->value) == type)
+    {
+        return true;
+    }
+    else {
+        reportError("Variable already declared: " + node->value, node);
+        return false;
+    }
+
+    return true;
+}
+
+// Check for function definitions
+void SemanticAnalyzer::clearTemp()
+{
+    temp->clear();
+}
+
+bool SemanticAnalyzer::checkVariableUsage(ASTNode* node, string type)
+{
+    if(!(scopeManager->lookup(node->value)))
+    {
+        reportError("Variable is not defined in this scope: " + node->value, node);
+        return false;
+    }else if(scopeManager->lookup(node->value) && type == "ANY")
+    {
+        return true;
+    }else if (scopeManager->lookup(node->value) && scopeManager->lookup(node->value) == type)
+    {
+        return true;
+    }
+    else if (type == "LOOKUP" && (scopeManager->lookup(node->value) == "NODE" ||
+                                  scopeManager->lookup(node->value) == "RELATIONSHIP" ||
+                                  scopeManager->lookup(node->value) == "MAP"))
+    {
+        return true;
+    }
+    else
+    {
+        reportError("Variable type mismatch: " + node->value, node);
+        return false;
+    }
+
+}
+
+// Report errors
+void SemanticAnalyzer::reportError(const std::string &message, ASTNode* node) {
+    std::cerr << "Error: " << message << std::endl;
```

Comment:
Can we use a logger here?

Reply:
sure sir, I will fix that.

Reply:
Ok, but there are some further changes required. Please check my new comments.